src/routes/(app)/admin[[hash=admin_hash]]/discount/[id]/+page.server.ts

```diff
@@ -46,30 +47,51 @@ export const actions = {
 			throw error(404, 'discount not found');
 		}
 
-		const data = await request.formData();
-
-		const { name, subscriptionIds, productIds, wholeCatalog, percentage, beginsAt, endsAt } = z
+		const formData = await request.formData();
+		const quantityPerProductRecord: Record<string, number> = {};
+		for (const [key, value] of formData.entries()) {
+			const match = key.match(/^quantityPerProduct\[(.+?)\]$/);
+			if (match) {
+				quantityPerProductRecord[match[1]] = Number(value);
+			}
+		}
+		const {
+			name,
+			subscriptionIds,
+			productIds,
+			wholeCatalog,
+			percentage,
+			beginsAt,
+			endsAt,
+			quantityPerProduct
+		} = z
 			.object({
 				name: z.string().min(1).max(MAX_NAME_LIMIT),
-				productIds: z.string().array(),
+				productIds: z.string().array().optional(),
 				subscriptionIds: z.string().array(),
-				percentage: z.string().regex(/^\d+(\.\d+)?$/),
+				percentage: z
+					.string()
+					.regex(/^\d+(\.\d+)?$/)
+					.optional(),
```

Comment:
It shouldn't be put as optional.


You parse one or the other, depending on `discount.mode`.

But currently on the frontend I can put all those as empty even for "percent" discounts